File: Stack/341_FlattenNestedListIterator.java
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */
public class NestedIterator implements Iterator<Integer> {
    private int index;
    private List<Integer> results;
    public NestedIterator(List<NestedInteger> nestedList) {
        results = new ArrayList<>();
        index = 0;
        init(nestedList);
    }

    private void init(List<NestedInteger> nestedList) {
        for (NestedInteger ni : nestedList) {
            if (ni.isInteger()) {
                results.add(ni.getInteger());
            } else {
                init(ni.getList());
            }
        }
    }

    @Override
    public Integer next() {
        return results.get(index++);
    }

    @Override
    public boolean hasNext() {
        return index < results.size();
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: Stack/341_FlattenNestedListIterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger(object):
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """

class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.result = []
        self.index = 0
        self.init_result(nestedList)

    def init_result(self, nestedList):
        for ele in nestedList:
            if ele.isInteger():
                self.result.append(ele.getInteger())
            else:
                newNestedList = ele.getList()
                self.init_result(newNestedList)

    def next(self):
        """
        :rtype: int
        """
        val = self.result[self.index]
        self.index += 1
        return val

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.index < len(self.result)

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())